src/emulator/eventarcEmulator.ts

```diff
@@ -0,0 +1,169 @@
+      for (const event of body.events) {
+        if (!event.type) {
+          res.sendStatus(400);
+          return;
+        }
+        logger.info(`Received custom event: ${JSON.stringify(event)}`);
+        this.triggerCustomEventFunction(channel, event);
+      }
+      res.sendStatus(200);
+    };
+
+    const dataMiddleware: express.RequestHandler = (req, _, next) => {
+      const chunks: Buffer[] = [];
+      req.on("data", (chunk: Buffer) => {
+        chunks.push(chunk);
+      });
+      req.on("end", () => {
+        (req as RequestWithRawBody).rawBody = Buffer.concat(chunks);
+        next();
+      });
+    };
+
+    const hub = express();
+    hub.post([registerTriggerRoute], dataMiddleware, registerTriggerHandler);
+    hub.post([publishEventsRoute], dataMiddleware, publishEventsHandler);
+    hub.all("*", (req, res) => {
+      logger.debug(`Eventarc emulator received unknown request at path ${req.path}`);
+      res.sendStatus(404);
+    });
+    return hub;
+  }
+
+  async triggerCustomEventFunction(channel: string, event: CloudEvent<any>) {
+    const functionsEmulator = EmulatorRegistry.get(Emulators.FUNCTIONS);
+    if (!functionsEmulator) {
+      logger.info("Functions emulator not found. This should not happen.");
+      return Promise.reject();
+    }
+    const key = `${event.type}-${channel}`;
+    const triggers = this.customEvents[key] || [];
+    return await Promise.all(
+      triggers
+        .filter(
+          (trigger) =>
+            !trigger.eventTrigger.eventFilters ||
+            this.matchesAll(event, trigger.eventTrigger.eventFilters)
+        )
+        .map((trigger) =>
+          api
+            .request(
+              "POST",
+              `/functions/projects/${trigger.projectId}/triggers/${trigger.triggerName}`,
+              {
+                origin: `http://${EmulatorRegistry.getInfoHostString(functionsEmulator.getInfo())}`,
+                data: JSON.stringify(event),
+                json: false,
+              }
+            )
+            .then(() => true)
+            .catch((err) => {
+              logger.error(
+                `Failed to trigger Functions emulator for ${trigger.triggerName}: ${err}`
+              );
+            })
+        )
+    );
+  }
+
+  private matchesAll(event: CloudEvent<any>, eventFilters: Record<string, string>): boolean {
+    return Object.entries(eventFilters).every(([key, value]) => {
+      let attr = event[key] ?? event.attributes[key];
+      if (typeof attr === "object" && !Array.isArray(attr)) {
+        attr = attr.ceTimestamp ?? attr.ceString;
+      }
+      return attr === value;
+    });
+  }
+
+  async start(): Promise<void> {
+    const { host, port } = this.getInfo();
+    const server = this.createHubServer().listen(port, host);
+    this.destroyServer = createDestroyer(server);
+    return Promise.resolve();
+  }
+
+  async connect(): Promise<void> {
+    // wip. what to do here?
```

Comment:
Probably nothing - in other emulators, we use this to set up any file watchers we need (ie to watch for changes to security rules in Storage/RTDB, or to watch for code changes in functions).

Can just replace this with `return Promise.resolve()`